[FLINK-10625] Documentation for MATCH_RECOGNIZE clause (#7070)
apache/flink: docs/dev/table/streaming/match_recognize.md

The schema of the output row will be following union of `{partitioning columns} + {measures columns}` in that particular order.

Example:

Query:
{% highlight sql %}
SELECT * 
FROM Ticker
    MATCH_RECOGNIZE(
        PARTITION BY symbol
        ORDER BY rowtime
        MEASURES
            FIRST(A.price) as startPrice
            LAST(A.price) as topPrice
            B.price as lastPrice
        PATTERN (A+ B)
        ONE ROW PER MATCH
        DEFINE
            A as A.price > LAST(A.price, 1) OR LAST(A.price, 1) IS NULL,
            B as B.price < LAST(A.price)
    )
{% endhighlight %}

for input:

{% highlight text %}
 symbol   tax   price.         rowtime
======== ===== ======== =====================
 XYZ      1     10       2018-09-17 10:00:02
 XYZ      2     12       2018-09-17 10:00:03
 XYZ      1     13       2018-09-17 10:00:04
 XYZ      2     11       2018-09-17 10:00:05
{% endhighlight %}

will produce:

{% highlight text %}
 symbol   startPrice   topPrice   lastPrice
======== ============ ========== ===========
 XYZ      10           13         11
{% endhighlight %}

Event stream navigation
------------------

### Pattern variable reference

> Pattern variable reference e.g. A.price describes a set of rows mapped so far to A plus current row, if we try to match current row to A. If the expression requires a single row e.g. A.price > 10, A.price selects the last value of the row.

Reviewer: Add an explanation sentences first before starting with a concrete example.

If no pattern variable is specified e.g. SUM(price) it references the default pattern variable which references all variables in the pattern, so it constitutes row set of all rows mapped so far plus current row.

Example:

{% highlight sql %}
...
PATTERN (A B+)
DEFINE
  A as A.price > 10,
  B as B.price > A.price AND SUM(price) < 100 AND SUM(B.price) < 80
{% endhighlight %}
  
By `{A.price}` we describe set of rows that the values in DEFINE clause are evaluated on. The rows are referenced by its number

> <table class="table table-bordered">

Reviewer: Use markdown syntax instead?

Author: I tried it, but I couldn't make it bordered and found it harder to read, moreover for the next table I control which rows can be wrapped. I think this is not supported in the markdown syntax as well, but I wanted to keep the same style for both tables.

  <thead>
    <tr>
      <th>no</th>
      <th>price</th>

>       <th>CLAS</th>

Reviewer: Can you explain this table a little bit more? What does `CLAS` mean?

      <th>{A.price}</th>
      <th>{B.price}</th>
      <th>{price}</th>
      <th>A.price</th>
      <th>B.price</th>
      <th>SUM(price)</th>
      <th>SUM(B.price)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>1</td>
      <td>10</td>
      <td>-&gt; A</td>
      <td>1</td>
      <td>-</td>
      <td>-</td>
      <td>10</td>
      <td>-</td>
      <td>-</td>
      <td>-</td>
    </tr>
    <tr>
      <td>2</td>
      <td>15</td>
      <td>-&gt; B</td>
      <td>1</td>
      <td>2</td>
      <td>1,2</td>
      <td>10</td>
      <td>15</td>
      <td>25</td>
      <td>15</td>
    </tr>
    <tr>
      <td>3</td>
      <td>20</td>
      <td>-&gt; B</td>
      <td>1</td>
      <td>2,3</td>

>       <td>1,2,3</td>

Reviewer: Use the actual prices instead of the row number?

Author: I didn't want to do that to differentiate it from actual evaluation to price in the next rows.

      <td>10</td>
      <td>20</td>
      <td>45</td>
      <td>35</td>
    </tr>
    <tr>
      <td>4</td>
      <td>31</td>
      <td>-&gt; B</td>
      <td>1</td>
      <td>2,3,4</td>
      <td>1,2,3,4</td>
      <td>10</td>
      <td>31</td>
      <td>76</td>
      <td>66</td>
    </tr>
    <tr>
      <td>5</td>
      <td>35</td>
      <td></td>
      <td>1</td>
      <td>2,3,4,5</td>
      <td>1,2,3,4,5</td>
      <td>10</td>
      <td>35</td>
      <td>111</td>
      <td>101</td>
    </tr>
  </tbody>
</table>

### Pattern variable indexing

> Apache Flink allows navigating within events that were mapped to a particular pattern variable with so called logical offsets. This can be expressed

Reviewer: Mention that this only makes sense when using `+`, `*` etc.

Author: Actually it makes sense in other cases as well:
* for the default pattern variable `LAST(price, 2)`, which indexes all events mapped so far 
* will make even more sense once we fully support `SUBSET`

Reviewer: You are right. Maybe you can add your example as well in this section.

* AFTER MATCH SKIP PAST LAST ROW

{% highlight text %}
 symbol   sumPrice        startTime              endTime
======== ========== ===================== =====================
 XYZ      26         2018-09-17 10:00:01   2018-09-17 10:00:03
 XYZ      17         2018-09-17 10:00:04   2018-09-17 10:00:04
{% endhighlight %}

* AFTER MATCH SKIP TO LAST A

{% highlight text %}
 symbol   sumPrice        startTime              endTime
======== ========== ===================== =====================
 XYZ      26         2018-09-17 10:00:01   2018-09-17 10:00:03
 XYZ      27         2018-09-17 10:00:03   2018-09-17 10:00:04
 XYZ      17         2018-09-17 10:00:04   2018-09-17 10:00:04
{% endhighlight %}

* AFTER MATCH SKIP TO NEXT ROW

{% highlight text %}
 symbol   sumPrice        startTime              endTime
======== ========== ===================== =====================
 XYZ      26         2018-09-17 10:00:01   2018-09-17 10:00:03
 XYZ      19         2018-09-17 10:00:02   2018-09-17 10:00:03
 XYZ      27         2018-09-17 10:00:03   2018-09-17 10:00:04
 XYZ      17         2018-09-17 10:00:04   2018-09-17 10:00:04
{% endhighlight %}

* AFTER MATCH SKIP TO FIRST A

> This combination will produce runtime exception, because one would try to always start new match where the

Reviewer: In general for this document, I would not use the word exception, but simply "is not supported".

Author: I wanted it to stand out from a situation when we throw exception while compiling query.
This can be compiled, but will throw exception during execution.

Reviewer: If your think we need to distinguish between runtime and non-runtime exceptions, we can leave it as it is. I was just wondering how often `exception` occurs in this document. It is obvious that unsupported things throw exceptions at some point.

  * reluctant optional quantifier - `PATTERN A??` only greedy optional quantifier is supported
* `ALL ROWS PER MATCH` output mode, which produces output row for every row that participated in creating a found match. This also means:
  * that the only supported semantic for `MEASURES` clause is `FINAL`
  * `CLASSIFIER` function, which returns the pattern variable that row was mapped to, is not yet supported
* `SUBSET` - which allows creating logical groups of pattern variables and using those groups in `DEFINE` and `MEASURES` clauses
* physical offsets - `PREV/NEXT`, which index all events seen rather than only those that were mapped to a pattern variable(as in [logical offsets](#pattern-variable-indexing) case)
* there is no support of aggregates yet, one cannot use aggregates in `MEASURES` nor `DEFINE` clauses
* user defined functions cannot be used within `MATCH_RECOGNIZE`
* `MATCH_RECOGNIZE` is supported only for SQL, there is no equivalent in the table API

### Controlling memory consumption

Memory consumption is an important aspect when writing `MATCH_RECOGNIZE` queries, as the space of potential of potential matches in built in a breadth first like manner.

> Having that in mind one must make sure that the pattern can finish(preferably with reasonable number of rows mapped to the match, as they have to fit into memory) e.g. it does not have a quantifier without

Reviewer: Can you give an example here?

A pattern variable `PRICE_DOWN` is defined as a row with a price that is smaller than the price of the last row that met the `PRICE_DOWN` condition. For the initial case or when there is no last row that met the `PRICE_DOWN` condition, the price of the row should be smaller than the price of the preceding row in the pattern (referenced by `START_ROW`).

A pattern variable `PRICE_UP` is defined as a row with a price that is larger than the price of the last row that met the `PRICE_DOWN` condition.

The query produces a summary row for each found period in which the price was constantly decreasing.

The exact representation of output rows is defined in `MEASURES` part of the query. The number of output rows is defined by the `ONE ROW PER MATCH` output mode.

{% highlight text %}
 symbol       start_tstamp       bottom_tstamp         end_tstamp
=========  ==================  ==================  ==================
ACME       01-APR-11 10:00:04  01-APR-11 10:00:07  01-APR-11 10:00:08
{% endhighlight %}

> The resulting row contains the found period of a decreasing price for a ticker that started at `01-APR-11 10:00:04` and

Reviewer 2: ```suggestion
The resulting row describes a period of falling prices that started at `01-APR-11 10:00:04` and
```

achieved the lowest price at `01-APR-11 10:00:07` that increased again at `01-APR-11 10:00:08`.

Partitioning
------------

It is possible to look for patterns in partitioned data, e.g., trends for a single ticker or a particular user. This can be expressed using the `PARTITION BY` clause. The clause is similar to using `GROUP BY` for aggregations.

<span class="label label-danger">Attention</span> It is highly advised to partition the incoming data because otherwise the `MATCH_RECOGNIZE` clause will be translated
into a non-parallel operator to ensure global ordering.

Order of Events
---------------

Apache Flink allows for searching for patterns based on time; either [processing time or event time](time_attributes.html).

> In case of event time, this assumption is very important because it enables to sort the events before they are passed to the internal pattern state machine. As a consequence, the

Reviewer 2: ```suggestion
In case of event time, the events are sorted before they are passed to the internal pattern state machine. As a consequence, the
```

produced output will be correct regardless of the order in which rows are appended to the table. Instead, the pattern is evaluated in the order specified by the time contained in each row.

The `MATCH_RECOGNIZE` clause assumes a [time attribute](time_attributes.html) with ascending ordering as the first argument to `ORDER BY` clause.

For the example `Ticker` table, a definition like `ORDER BY rowtime ASC, price DESC` is valid but `ORDER BY price, rowtime` or `ORDER BY rowtime DESC, price ASC` is not.

Define & Measures
-----------------

> The `DEFINE` and `MEASURES` keywords have similar meaning as `WHERE` and `SELECT` clauses in a simple SQL query.

Reviewer 2: ```suggestion
The `DEFINE` and `MEASURES` keywords have similar meanings to the `WHERE` and `SELECT` clauses in a simple SQL query.
```

> The `MEASURES` clause defines what will be included in the output of a matching pattern. It can project columns and defines expressions for evaluation.

Reviewer 2: ```suggestion
The `MEASURES` clause defines what will be included in the output of a matching pattern. It can project columns and define expressions for evaluation.
```

The number of produced rows depends on the [output mode](#output-mode) setting.

> The `DEFINE` clause allows to specify conditions that rows have to fulfill in order to be classified to a corresponding [pattern variable](#defining-pattern).

Reviewer 2: ```suggestion
The `DEFINE` clause specifies conditions that rows have to fulfill in order to be classified to a corresponding [pattern variable](#defining-pattern).
```

If a condition is not defined for a pattern variable, a default condition will be used which evaluates to `true` for every row.

For a more detailed explanation about expressions that can be used in those clauses, please have a look at the [event stream navigation](#pattern-navigation) section.

Defining a Pattern
------------------

The `MATCH_RECOGNIZE` clause allows users to search for patterns in event streams using a powerful and expressive syntax
that is somewhat similar to the widespread regular expression syntax.

Every pattern is constructed from basic building blocks, called _pattern variables_, to which operators (quantifiers and other modifiers) can be applied. The whole pattern must be enclosed in
brackets.

An example pattern could like:

{% highlight sql %}

> PATTERN (A B+ C*? D)

Reviewer 2: I gather (after reading further) that ? has been overloaded. But when I first read this example the ? seemed like a mistake -- I thought C*? would be equivalent to C*. I suggest removing the ? from this example, as I found it quite distracting.

FROM Ticker
    MATCH_RECOGNIZE(
        PARTITION BY symbol
        ORDER BY rowtime
        MEASURES
            FIRST(A.price) AS startPrice
            LAST(A.price) AS topPrice
            B.price AS lastPrice
        PATTERN (A+ B)
        ONE ROW PER MATCH
        DEFINE
            A AS LAST(A.price, 1) IS NULL OR A.price > LAST(A.price, 1),
            B AS B.price < LAST(A.price)
    )
{% endhighlight %}

For the following input rows:

{% highlight text %}
 symbol   tax   price          rowtime
======== ===== ======== =====================
 XYZ      1     10       2018-09-17 10:00:02
 XYZ      2     12       2018-09-17 10:00:03
 XYZ      1     13       2018-09-17 10:00:04
 XYZ      2     11       2018-09-17 10:00:05
{% endhighlight %}

The query will produce the following output:

{% highlight text %}
 symbol   startPrice   topPrice   lastPrice
======== ============ ========== ===========
 XYZ      10           13         11
{% endhighlight %}

The pattern recognition is partitioned by the `symbol` column. Even though not explicitly mentioned in the `MEASURES` clause, the partitioned column is added at the beginning of the result.

Pattern Navigation
------------------

> The `DEFINE` and `MEASURE` clauses allow for navigating within a list of (potentially) matching rows to a pattern.

Reviewer 2: ```suggestion
The `DEFINE` and `MEASURE` clauses allow for navigating within the list of rows that (potentially) match a pattern.
```

This section discusses this navigation for declaring conditions or producing output results.

### Pattern Variable Referencing

> A _pattern variable reference_ allows to reference a set of rows mapped to a particular pattern variable in

Reviewer 2: A _pattern variable reference_ allows a set of rows mapped to a particular pattern variable in `DEFINE` or `MEASURE` clauses to be referenced.

    </tr>
    <tr>
      <td>#3</td>
      <td>20</td>
      <td>-&gt; B</td>
      <td>#1</td>
      <td>#2, #3</td>
      <td>#1, #2, #3</td>
      <td>10</td>
      <td>20</td>
      <td>45</td>
      <td>35</td>
    </tr>
    <tr>
      <td>#4</td>
      <td>31</td>
      <td>-&gt; B</td>
      <td>#1</td>
      <td>#2, #3, #4</td>
      <td>#1, #2, #3, #4</td>
      <td>10</td>
      <td>31</td>
      <td>76</td>
      <td>66</td>
    </tr>
    <tr>
      <td>#5</td>
      <td>35</td>
      <td></td>
      <td>#1</td>
      <td>#2, #3, #4, #5</td>
      <td>#1, #2, #3, #4, #5</td>
      <td>10</td>
      <td>35</td>
      <td>111</td>
      <td>101</td>
    </tr>
  </tbody>
</table>

As can be seen in the table, the first row is mapped to pattern variable `A` and subsequent rows are mapped to pattern variable `B`. However, the last row does not fulfill the `B` condition because the sum over all mapped rows `SUM(price)` and the sum over all rows in `B` exceed the specified thresholds.

<span class="label label-danger">Attention</span> Please note that aggregations such as `SUM` are not supported yet. They are only used for explanation here.

### Logical Offsets

> _Logical offsets_ allow to navigate within events that were mapped to a particular pattern variable. This can be expressed

Reviewer 2: ```suggestion
_Logical offsets_ enable navigation within the events that were mapped to a particular pattern variable. This can be expressed
```

      <td>15</td>
      <td><code>LAST(price, 1)</code> is evaluated as the price of the row mapped to the <code>B</code> variable.</td>
    </tr>
  </tbody>
</table>

If the second row did not map to the `B` variable, we would have the following results:

<table class="table table-bordered">
  <thead>
    <tr>
        <th style="white-space:nowrap">price</th>
        <th style="white-space:nowrap">Classifier</th>
        <th style="white-space:nowrap">LAST(price, 1)</th>
        <th>Comment</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>10</td>
      <td>-&gt; A</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>20</td>
      <td>-&gt; C</td>
      <td>10</td>
      <td><code>LAST(price, 1)</code> is evaluated as the price of the row mapped to the <code>A</code> variable.</td>
    </tr>
  </tbody>
</table>

It is also possible to use multiple pattern variable references in the first argument of the `FIRST/LAST` functions. This way, one can write an expression that accesses multiple columns.
However, all of them must use the same pattern variable. In other words, the value of the `LAST`/`FIRST` function must be computed in a single row.

> Thus, it is possible to use `LAST(A.price * A.tax)` but an expression like `LAST(A.price * B.tax)` is not allowed.

Reviewer 2: ```suggestion
Thus, it is possible to use `LAST(A.price * A.tax)`, but an expression like `LAST(A.price * B.tax)` is not allowed.
```

After Match Strategy
--------------------

The `AFTER MATCH SKIP` clause specifies where to start a new matching procedure after a complete match was found.

> There are five different strategies:

Reviewer 2: I only see four, not five.

 XYZ      26         2018-09-17 10:00:01   2018-09-17 10:00:04
 XYZ      15         2018-09-17 10:00:03   2018-09-17 10:00:05
 XYZ      22         2018-09-17 10:00:04   2018-09-17 10:00:06
 XYZ      17         2018-09-17 10:00:05   2018-09-17 10:00:06
{% endhighlight %}

Again, the first result matched against the rows #1, #2, #3, #4.

Compared to the previous strategy, the next match includes only row #3 (mapped to `A`) again for the next matching. Therefore, the second result matched against the rows #3, #4, #5.

The third result matched against the rows #4, #5, #6.

The last result matched against the rows #5, #6.

##### `AFTER MATCH SKIP TO FIRST A`

This combination will produce a runtime exception because one would always try to start a new match where the
last one started. This would produce an infinite loop and, thus, is prohibited.

One has to keep in mind that in case of the `SKIP TO FIRST/LAST variable`strategy it might be possible that there are no rows mapped to that
variable (e.g. for pattern `A*`). In such cases, a runtime exception will be thrown as the standard requires a valid row to continue the
matching.

### Controlling Memory Consumption

> Memory consumption is an important aspect when writing `MATCH_RECOGNIZE` queries as the space of potential matches is built in a breadth-first like manner.

Reviewer 2: ```suggestion
Memory consumption is an important consideration when writing `MATCH_RECOGNIZE` queries, as the space of potential matches is built in a breadth-first-like manner.
```

For example, the pattern must not have a quantifier without an upper limit that accepts every single row. Such a pattern could look like this:

{% highlight sql %}
PATTERN (A B+ C)
DEFINE
  A as A.price > 10,
  C as C.price > 20
{% endhighlight %}

The query will map every incoming row to the `B` variable and thus will never finish. This query could be fixed, e.g., by negating the condition for `C`:

{% highlight sql %}
PATTERN (A B+ C)
DEFINE
  A as A.price > 10,
  B as B.price <= 20,
  C as C.price > 20
{% endhighlight %}

Or by using the [reluctant quantifier](#greedy--reluctant-quantifiers):

{% highlight sql %}
PATTERN (A B+? C)
DEFINE
  A as A.price > 10,
  C as C.price > 20
{% endhighlight %}

<span class="label label-danger">Attention</span> Please note that the `MATCH_RECOGNIZE` clause does not use a configured [state retention time](query_configuration.html#idle-state-retention-time). As of now, there is also no possibility to define a time restriction on the pattern to finish because there is no such possibility in the SQL standard. The community is in the process of designing a proper syntax for that
feature right now.

Known Limitations
-----------------

> `MATCH_RECOGNIZE` clause is still an ongoing effort and therefore some SQL standard features are not supported yet.

Reviewer 2: ```suggestion
Flink's implementation of the `MATCH_RECOGNIZE` clause is an ongoing effort, and some features of the SQL standard are not yet supported.
```

> The list of such features includes:

Reviewer 2: ```suggestion
Unsupported features include:
```

* Pattern expressions:
  * Pattern groups - this means that e.g. quantifiers can not be applied to a subsequence of the pattern. Thus, `(A (B C)+)` is not a valid pattern.
  * Alterations - patterns like `PATTERN((A B | C D) E)`, which means that either a subsequence `A B` or `C D` has to be found before looking for the `E` row.
  * `PERMUTE` operator - which is equivalent to all permutations of variables that it was applied to e.g. `PATTERN (PERMUTE (A, B, C))` = `PATTERN (A B C | A C B | B A C | B C A | C A B | C B A)`.
  * Anchors - `^, $`, which denote beginning/end of a partition, those do not make sense in the streaming context and will not be supported.

>   * Exclusion - `PATTERN ({- A -} B)` meaning that `A` will be looked for but will not participate in the output. This works only for the `ALL ROWS PER MATCH` mode.

Reviewer 2: Are you saying that exclusion only makes sense in the ALL ROWS PER MATCH mode, and Flink doesn't support exclusion because Flink doesn't support ALL ROWS PER MATCH -- or that Flink currently only supports exclusion in ALL ROWS PER MATCH mode?

Reviewer 2: I see this becomes clearer about two lines later ...

  * Reluctant optional quantifier - `PATTERN A??` only the greedy optional quantifier is supported.
* `ALL ROWS PER MATCH` output mode - which produces an output row for every row that participated in the creation of a found match. This also means:
  * that the only supported semantic for the `MEASURES` clause is `FINAL`
  * `CLASSIFIER` function, which returns the pattern variable that a row was mapped to, is not yet supported.
* `SUBSET` - which allows creating logical groups of pattern variables and using those groups in the `DEFINE` and `MEASURES` clauses.
* Physical offsets - `PREV/NEXT`, which indexes all events seen rather than only those that were mapped to a pattern variable(as in [logical offsets](#logical-offsets) case).

> * Extracting time attributes - there is no possibility to get a time attribute for subsequent time-based operation yet.

Reviewer 2: ```suggestion
* Extracting time attributes - there is currently no possibility to get a time attribute for subsequent time-based operations.
```

> * Aggregates - one cannot use aggregates in `MEASURES` nor `DEFINE` clauses yet.

Reviewer 2: ```suggestion
* Aggregates - one cannot use aggregates in `MEASURES` nor `DEFINE` clauses.
```
